specs/phase1/shard-fork-choice.md

```diff
@@ -0,0 +1,147 @@
+    block_states: Dict[Root, ShardState] = field(default_factory=dict)
+```
+
+#### `get_forkchoice_shard_store`
+
+```python
+def get_forkchoice_shard_store(anchor_state: BeaconState, shard: Shard) -> ShardStore:
+    return ShardStore(
+        shard=shard,
+        blocks={anchor_state.shard_states[shard].latest_block_root: ShardBlock(slot=anchor_state.slot)},
+        block_states={anchor_state.shard_states[shard].latest_block_root: anchor_state.copy().shard_states[shard]},
+    )
+```
+
+#### `get_shard_latest_attesting_balance`
+
+```python
+def get_shard_latest_attesting_balance(store: Store, shard_store: ShardStore, root: Root) -> Gwei:
+    state = store.checkpoint_states[store.justified_checkpoint]
+    active_indices = get_active_validator_indices(state, get_current_epoch(state))
+    return Gwei(sum(
+        state.validators[i].effective_balance for i in active_indices
+        if (
+            i in store.latest_messages and
+            store.latest_messages[i].shard == shard_store.shard and
+            get_shard_ancestor(
+                store, shard_store, store.latest_messages[i].root, shard_store.blocks[root].slot
+            ) == root
+        )
+    ))
+```
+
+#### `get_shard_head`
+
+```python
+def get_shard_head(store: Store, shard_store: ShardStore) -> Root:
+    # Execute the LMD-GHOST fork choice
+    shard_blocks = shard_store.blocks
+    head_beacon_root = get_head(store)
+    head_shard_state = store.block_states[head_beacon_root].shard_states[shard_store.shard]
+    head_shard_root = head_shard_state.latest_block_root
+    while True:
+        # Find the valid child block roots
+        children = [
+            root for root in shard_store.blocks.keys()
+            if (
+                shard_blocks[root].shard_parent_root == head_shard_root
+                and shard_blocks[root].slot > head_shard_state.slot
+            )
+        ]
+        if len(children) == 0:
+            return head_shard_root
+        # Sort by latest attesting balance with ties broken lexicographically
+        head_shard_root = max(
+            children, key=lambda root: (get_shard_latest_attesting_balance(store, shard_store, root), root)
+        )
+```
+
+#### `get_shard_ancestor`
+
+```python
+def get_shard_ancestor(store: Store, shard_store: ShardStore, root: Root, slot: Slot) -> Root:
+    block = shard_store.blocks[root]
+    if block.slot > slot:
+        return get_shard_ancestor(store, shard_store, block.shard_parent_root, slot)
+    elif block.slot == slot:
+        return root
+    else:
+        # root is older than queried slot, thus a skip slot. Return earliest root prior to slot
+        return root
+```
+
+### Handlers
+
+#### `on_shard_block`
+
+```python
+def on_shard_block(store: Store, shard_store: ShardStore, signed_shard_block: SignedShardBlock) -> None:
+    shard_block = signed_shard_block.message
+    shard = shard_store.shard
+    # 1. Check shard parent exists
```

Comment:
Add `assert shard_block.shard == shard` ?

Reply:
I thought about this and assumed that since the client would be the one piping the shard block into this function with the associated `shard_store` that this base mapping already occurred.

That said, I suppose there is not hurt in double checking here

Reply:
Fair point. I'll still add the check on Prysm just to be extra defensive in good way

Reply:
I'm happy for it to go in.

Another way we can design this is to have `shard_stores` be a map of shard -> shard_store. And `on_shard_block` actually selects the `shard_store` to use.

Curious what @hwwhww feels about the options here

Reply:
I see that.